File: 078_ImplementQueueUsingStacks.cpp
// Topic: Implement Queue using Stacks
// Description: Implement a queue with push, pop, peek, and empty using two stacks.
// Approach: Use two stacks, 'in' for pushing and 'out' for popping. Amortized O(1) per operation.

#include <bits/stdc++.h>
using namespace std;

class MyQueue
{
    stack<int> in, out;
    void moveInToOut()
    {
        if (out.empty())
        {
            while (!in.empty())
            {
                out.push(in.top());
                in.pop();
            }
        }
    }

   public:
    MyQueue() {}
    void push(int x) { in.push(x); }
    int pop()
    {
        moveInToOut();
        int v = out.top();
        out.pop();
        return v;
    }
    int peek()
    {
        moveInToOut();
        return out.top();
    }
    bool empty() { return in.empty() && out.empty(); }
};

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    vector<vector<pair<string, int>>> tests = {
        {{"push", 1}, {"push", 2}, {"push", 3}, {"pop", 0}, {"pop", 0}, {"pop", 0}},
        {{"push", 10}, {"peek", 0}, {"pop", 0}, {"empty", 0}},
        {{"push", 5}, {"pop", 0}}};

    for (int t = 0; t < (int)tests.size(); ++t)
    {
        MyQueue q;
        cout << "Test " << (t + 1) << ": ";
        for (auto& op : tests[t])
        {
            if (op.first == "push")
                q.push(op.second);
            else if (op.first == "pop")
                cout << q.pop() << ' ';
            else if (op.first == "peek")
                cout << q.peek() << ' ';
            else if (op.first == "empty")
                cout << (q.empty() ? 1 : 0) << ' ';
        }
        cout << '\n';
    }

    return 0;
}

/*
Time Complexity: Amortized O(1) per operation
Space Complexity: O(n)
*/

File: 078_ImplementQueueUsingStacks.py
# Topic: Implement Queue using Stacks
# Description: Implement a queue with push, pop, peek, and empty using two stacks.
# Approach: Use two stacks, 'in_stack' for pushing and 'out_stack' for popping. Amortized O(1) per operation.


class MyQueue:
    def __init__(self):
        self.in_stack = []
        self.out_stack = []

    def _move(self):
        if not self.out_stack:
            while self.in_stack:
                self.out_stack.append(self.in_stack.pop())

    def push(self, x: int) -> None:
        self.in_stack.append(x)

    def pop(self) -> int:
        self._move()
        return self.out_stack.pop()

    def peek(self) -> int:
        self._move()
        return self.out_stack[-1]

    def empty(self) -> bool:
        return not self.in_stack and not self.out_stack


if __name__ == "__main__":
    q = MyQueue()
    q.push(1)
    q.push(2)
    print(f"peek -> {q.peek()}")   # 1
    print(f"pop -> {q.pop()}")     # 1
    print(f"empty -> {q.empty()}") # False

"""
Time Complexity: Amortized O(1) per operation
Space Complexity: O(n)
"""
